plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +474,41 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
+    def process_complex_args(vars):
+        ret_out = []
+        if isinstance(vars, dict):
+            for k, v in vars.items():
+                if isinstance(v, dict):
+                    ret_out.append('{0}={{{1}}}'.format(k, process_complex_args(v)))
+                if isinstance(v, list):
+                    ret_out.append("{0}=[{1}]".format(k, process_complex_args(v)))
+                if isinstance(v, (integer_types, float, str, bool)):
+                    ret_out.append('{0}={1}'.format(k, json.dumps(v)))
+        if isinstance(vars, list):
+            l_out = []
+            for item in vars:
+                if isinstance(item, dict):
+                    l_out.append("{{{0}}}".format(process_complex_args(item)))
+                elif isinstance(item, (str, integer_types, float, bool)):
+                    l_out.append(json.dumps(item))
+                else:
+                    module.fail_json(msg="List can contain, dictionaries, strings, integer_types, boolean and float.")
+            ret_out.append("[{0}]".format(",".join(l_out)))
+        return ",".join(ret_out)
+
     variables_args = []
     for k, v in variables.items():
-        variables_args.extend([
-            '-var',
-            '{0}={1}'.format(k, v)
-        ])
+        if isinstance(v, (dict, list)):
+            variables_args.extend([
+                '-var',
+                '{0}={1}'.format(k, process_complex_args(v))
+            ])
+        else:
+            variables_args.extend([
+                '-var',
+                '{0}={1}'.format(k, v)
```

Comment:
Does this work for booleans? They will be `True` or `False` this way.

Reply:
This was from the original code segment, my test cases didn't have a top level boolean, and I'm sure it wouldn't work. I'm tempted to separate value formatting to a separate function. That way for the toplevel and well as nested type outputs can be maintained uniformly. 

I'll work on it shortly.